### ISSUE TITLE
Missing Batch_Normalization

### ISSUE THREAD

**Issue by tcwicks** [SciSharp/TensorFlow.NET#510]
Batch normalization does not appear to be implemented. I've looked for it in all the namespaces including Keras. There is a stub with no implementation though. How can we get batch normalization please.

**Comment by Oceania2018** [comment]
Can you provide the python doc?

**Comment by tcwicks** [comment]
I am extremely sorry for wasting your time. I searched the master and it looks like its now implemented here:

tf.layers.batch_normalization

https://github.com/SciSharp/TensorFlow.NET/blob/998a9ac4d248e029276b280e8772bf517c18c5ba/src/TensorFlowNET.Core/APIs/tf.layers.cs

Which is a port of:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/layers/normalization.py

Thankyou so very much.

**Comment by tcwicks** [comment]
Trying to use tf.layers.batch_normalization throws the following exception:

`
   System.NullReferenceException: Object reference not set to an instance of an object.
   at Tensorflow.tensor_util._ConstantValue(Tensor tensor, Boolean partial)
   at Tensorflow.Framework.smart_module.smart_constant_value(Tensor pred)
   at Tensorflow.Framework.smart_module.smart_cond[T](Tensor pred, Func`1 true_fn, Func`1 false_fn, String name)
   at Tensorflow.Keras.Layers.BatchNormalization._fused_batch_norm(Tensor inputs, Tensor training)
   at Tensorflow.Keras.Layers.BatchNormalization.call(Tensor inputs, Tensor training, Tensor state)
   at Tensorflow.Keras.Layers.Layer.<>c__DisplayClass23_0.<__call__>b__0(NameScope <p0>)
   at Tensorflow.Binding.tf_with[T](T py, Action`1 action)
   at Tensorflow.Keras.Layers.Layer.__call__(Tensor[] inputs, Tensor training, Tensor state, VariableScope scope)
   at Tensorflow.Layers.Layer.<>c__DisplayClass8_0.<__call__>b__0(variable_scope scope2)
   at Tensorflow.Binding.tf_with[T](T py, Action`1 action)
   at Tensorflow.Layers.Layer.__call__(Tensor inputs, Tensor training, Tensor state, VariableScope scope)
   at Tensorflow.Layers.Layer.apply(Tensor inputs, Tensor training)
   at Tensorflow.tensorflow.layers_internal.batch_normalization(Tensor inputs, Int32 axis, Single momentum, Single epsilon, Boolean center, Boolean scale, IInitializer beta_initializer, IInitializer gamma_initializer, IInitializer moving_mean_initializer, IInitializer moving_variance_initializer, Tensor training, Boolean trainable, String name, Boolean renorm, Single renorm_momentum)`

**Comment by Oceania2018** [comment]
Please provide a small piece UnitTest, it will make us work easier.

**Comment by tcwicks** [comment]
Invocation of tf.layers.batch_normalization fails at invocation.

`    // Just invoking tf.layers.batch_normalization fails

    Tensor Input = tf.placeholder(tf.float32, shape: new int[2] { -1, 16 }, name: "Input");

    Tensor Layer = FullyConnected(Input, 16, _Name: @"Layer", Trainable: true);

    Layer = tf.sigmoid(Layer);
    
    //This line throws an exception
    Layer = tf.layers.batch_normalization(Layer);
`

**Comment by tcwicks** [comment]
This is the utility function used in above code:

`
        //This is the method body for Fully Connected function call above

        public static Tensor FullyConnected(Tensor _Input, int _Num_Units, TF_DataType dataType = TF_DataType.TF_FLOAT, bool Trainable = true,
            float StandardDeviation = 0.01f, string _Name = null)
        {
            int in_dim = _Input.shape[1];
            string WeightsName;
            string BiasName;
            WeightsName = ConcatIfNotNullOrEmptyElseNull(_Name, @"_Weights");
            BiasName = ConcatIfNotNullOrEmptyElseNull(_Name, @"_Bias");

            if (in_dim == 0)
            {
                in_dim = -1;
            }
            RefVariable Weights;
            RefVariable Bias;
            if (Trainable)
            {
                Weights = tf.get_variable(WeightsName,
                    dtype: dataType,
                    shape: new int[2] { in_dim, _Num_Units },
                    trainable: true,
                    initializer: tf.truncated_normal_initializer(dtype: TF_DataType.TF_FLOAT),
                    validate_shape: false);


                Bias = tf.get_variable(BiasName,
                    dtype: dataType,
                    shape: new int[1] { _Num_Units },
                    trainable: true,
                    initializer: tf.truncated_normal_initializer(dtype: TF_DataType.TF_FLOAT),
                    validate_shape: false);
            }
            else
            {
                Weights = tf.get_variable(WeightsName,
                    dtype: dataType,
                    shape: new int[2] { in_dim, _Num_Units },
                    trainable: false,
                    initializer: tf.truncated_normal_initializer(dtype: TF_DataType.TF_FLOAT),
                    validate_shape: false);

                Bias = tf.get_variable(BiasName,
                    dtype: dataType,
                    shape: new int[1] { _Num_Units },
                    trainable: false,
                    initializer: tf.truncated_normal_initializer(dtype: TF_DataType.TF_FLOAT),
                    validate_shape: false);
            }

            Tensor layer = tf.matmul(_Input, Weights) + Bias;
            return layer;
        }
`

**Comment by dje-dev** [comment]
I am also trying to use batch_normalization and found the same Exception shown above. However I discovered that if a Tensor is provided to the training argument, this Exception is avoided. 

Unfortunately another Exception seems to then be thrown subsequently (at global variable initialization) that seems linked to batch_normalization, and have not been able to resolve this.

**Comment by Oceania2018** [comment]
@dje-dev @tcwicks `BatchNormalization` is implemented, check the [example](https://github.com/SciSharp/SciSharp-Stack-Examples/blob/4c814d6f6edd529c3637fe1dfb13a8b2ce6c27e1/src/TensorFlowNET.Examples/ImageProcessing/ToyResNet.cs#L60).
